File: cpp/src/domain/domain.cpp
#include "dhquant/domain.h"

namespace dhquant {

std::string_view to_string(const InstrumentType value) noexcept {
  switch (value) {
  case InstrumentType::kUnknown:
    return "unknown";
  case InstrumentType::kStock:
    return "stock";
  case InstrumentType::kEtf:
    return "etf";
  case InstrumentType::kIndex:
    return "index";
  }

  return "unknown";
}

std::string_view to_string(const Side value) noexcept {
  switch (value) {
  case Side::kUnknown:
    return "unknown";
  case Side::kBuy:
    return "buy";
  case Side::kSell:
    return "sell";
  }

  return "unknown";
}

std::string_view to_string(const Offset value) noexcept {
  switch (value) {
  case Offset::kUnknown:
    return "unknown";
  case Offset::kOpen:
    return "open";
  case Offset::kClose:
    return "close";
  case Offset::kCloseToday:
    return "close_today";
  case Offset::kCloseYesterday:
    return "close_yesterday";
  }

  return "unknown";
}

std::string_view to_string(const OrderType value) noexcept {
  switch (value) {
  case OrderType::kUnknown:
    return "unknown";
  case OrderType::kLimit:
    return "limit";
  case OrderType::kMarket:
    return "market";
  }

  return "unknown";
}

std::string_view to_string(const OrderStatus value) noexcept {
  switch (value) {
  case OrderStatus::kUnknown:
    return "unknown";
  case OrderStatus::kPendingNew:
    return "pending_new";
  case OrderStatus::kNew:
    return "new";
  case OrderStatus::kPartiallyFilled:
    return "partially_filled";
  case OrderStatus::kFilled:
    return "filled";
  case OrderStatus::kCancelPending:
    return "cancel_pending";
  case OrderStatus::kCancelled:
    return "cancelled";
  case OrderStatus::kRejected:
    return "rejected";
  }

  return "unknown";
}

std::string_view to_string(const RejectReason value) noexcept {
  switch (value) {
  case RejectReason::kNone:
    return "none";
  case RejectReason::kInvalidInstrument:
    return "invalid_instrument";
  case RejectReason::kInvalidPrice:
    return "invalid_price";
  case RejectReason::kInvalidQuantity:
    return "invalid_quantity";
  case RejectReason::kRiskRejected:
    return "risk_rejected";
  case RejectReason::kGatewayRejected:
    return "gateway_rejected";
  case RejectReason::kUnknown:
    return "unknown";
  }

  return "unknown";
}

std::string_view to_string(const RuntimeMode value) noexcept {
  switch (value) {
  case RuntimeMode::kUnknown:
    return "unknown";
  case RuntimeMode::kBacktest:
    return "backtest";
  case RuntimeMode::kPaper:
    return "paper";
  case RuntimeMode::kLive:
    return "live";
  }

  return "unknown";
}

bool is_terminal(const OrderStatus status) noexcept {
  return status == OrderStatus::kFilled || status == OrderStatus::kCancelled ||
         status == OrderStatus::kRejected;
}

} // namespace dhquant

File: python/dhquant/domain.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class InstrumentType(str, Enum):
    UNKNOWN = "unknown"
    STOCK = "stock"
    ETF = "etf"
    INDEX = "index"


class Side(str, Enum):
    UNKNOWN = "unknown"
    BUY = "buy"
    SELL = "sell"


class Offset(str, Enum):
    UNKNOWN = "unknown"
    OPEN = "open"
    CLOSE = "close"
    CLOSE_TODAY = "close_today"
    CLOSE_YESTERDAY = "close_yesterday"


class OrderType(str, Enum):
    UNKNOWN = "unknown"
    LIMIT = "limit"
    MARKET = "market"


class OrderStatus(str, Enum):
    UNKNOWN = "unknown"
    PENDING_NEW = "pending_new"
    NEW = "new"
    PARTIALLY_FILLED = "partially_filled"
    FILLED = "filled"
    CANCEL_PENDING = "cancel_pending"
    CANCELLED = "cancelled"
    REJECTED = "rejected"


class RejectReason(str, Enum):
    NONE = "none"
    INVALID_INSTRUMENT = "invalid_instrument"
    INVALID_PRICE = "invalid_price"
    INVALID_QUANTITY = "invalid_quantity"
    RISK_REJECTED = "risk_rejected"
    GATEWAY_REJECTED = "gateway_rejected"
    UNKNOWN = "unknown"


class RuntimeMode(str, Enum):
    UNKNOWN = "unknown"
    BACKTEST = "backtest"
    PAPER = "paper"
    LIVE = "live"


@dataclass
class PriceLevel:
    price: float = 0.0
    quantity: int = 0


@dataclass
class Instrument:
    instrument_id: str = ""
    exchange: str = ""
    symbol: str = ""
    instrument_type: InstrumentType = InstrumentType.UNKNOWN
    lot_size: int = 0
    price_tick: float = 0.0
    currency: str = "CNY"


@dataclass
class Tick:
    instrument_id: str = ""
    ts_event: int = 0
    ts_process: int = 0
    last_price: float = 0.0
    last_quantity: int = 0
    bids: list[PriceLevel] = field(default_factory=list)
    asks: list[PriceLevel] = field(default_factory=list)


@dataclass
class Bar:
    instrument_id: str = ""
    ts_event: int = 0
    ts_process: int = 0
    open: float = 0.0
    high: float = 0.0
    low: float = 0.0
    close: float = 0.0
    volume: int = 0
    turnover: float = 0.0


@dataclass
class OrderIntent:
    instrument_id: str = ""
    side: Side = Side.UNKNOWN
    offset: Offset = Offset.UNKNOWN
    order_type: OrderType = OrderType.UNKNOWN
    quantity: int = 0
    price: float = 0.0


@dataclass
class Order:
    session_id: str = ""
    order_id: str = ""
    instrument_id: str = ""
    side: Side = Side.UNKNOWN
    offset: Offset = Offset.UNKNOWN
    order_type: OrderType = OrderType.UNKNOWN
    status: OrderStatus = OrderStatus.UNKNOWN
    reject_reason: RejectReason = RejectReason.NONE
    quantity: int = 0
    filled_quantity: int = 0
    price: float = 0.0
    average_fill_price: float = 0.0
    ts_event: int = 0
    ts_process: int = 0


@dataclass
class Trade:
    session_id: str = ""
    order_id: str = ""
    trade_id: str = ""
    instrument_id: str = ""
    side: Side = Side.UNKNOWN
    offset: Offset = Offset.UNKNOWN
    fill_quantity: int = 0
    fill_price: float = 0.0
    commission: float = 0.0
    ts_event: int = 0
    ts_process: int = 0


@dataclass
class RiskEvent:
    order_id: str = ""
    instrument_id: str = ""
    passed: bool = False
    reason: RejectReason = RejectReason.NONE
    detail: str = ""
    ts_event: int = 0
    ts_process: int = 0


@dataclass
class Position:
    instrument_id: str = ""
    quantity_total: int = 0
    quantity_available: int = 0
    average_price: float = 0.0
    market_value: float = 0.0


@dataclass
class Account:
    account_id: str = ""
    currency: str = "CNY"
    cash: float = 0.0
    frozen_cash: float = 0.0
    equity: float = 0.0


@dataclass
class PortfolioSnapshot:
    snapshot_id: str = ""
    account_id: str = ""
    account_state: Account = field(default_factory=Account)
    positions: list[Position] = field(default_factory=list)
    ts_snapshot: int = 0


def is_terminal(status: OrderStatus) -> bool:
    return status in {
        OrderStatus.FILLED,
        OrderStatus.CANCELLED,
        OrderStatus.REJECTED,
    }
